### DISCUSSION TITLE
Question about pcm-iio measurement

### DISCUSSION THREAD

**Discussion by sakaia** [opcm/pcm#392] [General]
Question about pcm-iio measurement

1) Where is this traffic measured on? I want to see the throughput between device A and device B 
   Is the documentation available ?I am just looking following document. 
       https://www.intel.com/content/www/us/en/developer/articles/technical/utilizing-the-intel-xeon-processor-scalable-family-iio-performance-monitoring-events.html
2) pcm-iio duration time is 3 second. So the throughput per  sec should devide by 3?

```
 _____________________________ ________________ _______________ _______________ ________________ _________ __________ ______________ ____________ ____________ ____________ _____________
|IIO Stack 3 - PCIe2          |IB write (bytes)|IB read (bytes)|OB read (bytes)|OB write (bytes)|IOTLB Hit|IOTLB Miss|VT-d CTXT Miss|VT-d L1 Miss|VT-d L2 Miss|VT-d L3 Miss|VT-d Mem Read|
|_____________________________|________________|_______________|_______________|________________|_________|__________|______________|____________|____________|____________|_____________|
| Part0 (1st x16/x8/x4)       |1495 M          |1189 M         | 143 K         |1225 K          |  19 M   | 192      |   0          |   0        |   0        |   0        | 553         |
| Part1 (2nd x4)              |   0            |   0           |   0           |   0            |
| Part2 (2nd x8/3rd x4)       |   0            |   0           |   0           |   0            |
| Part3 (4th x4)              |   0            |   0           |   0           |   0            |
|_____________________________|________________|_______________|_______________|________________|_________|__________|______________|____________|____________|____________|_____________|
```

**Comment by ilia-kurakin** [comment]
Hello,
Traffic is measured in the IIO PMU, the doc you're referring to is the good one. Also refer to uncore performance monitoring guides, e.g. that one is for Skylake servers: https://cdrdv2.intel.com/v1/dl/getContent/671389?explicitVersion=true&wapkw=xeon%20scalable%20uncore

>>> I want to see the throughput between device A and device B

Do you mean your devices perform "peer to peer" accesses? Note that pcm-iio shows only device to main memory (inbound, "IB") and cores to devices (outbound, "OB") traffic.

If you want to see per-sec counts, specify sampling interval like this:
pcm-iio.x 1

**Comment by sakaia** [comment]
@ilia-kurakin  Thank you for description
Is there any peer-to-peer traffic monitoring tool?

**Comment by ilia-kurakin** [comment]
@sakaia there's no tool at the moment that implements p2p bandwidth metrics. However you can collect raw IIO events (e.g. with pcm-raw):
DATA_REQ_OF_CPU.PEER_READ.PART{x}, DATA_REQ_OF_CPU.PEER_WRITE.PART{x} - count every 4 bytes of inbound p2p data
DATA_REQ_BY_CPU.PEER_READ.PART{x}, DATA_REQ_BY_CPU.PEER_WRITE.PART{x} - count every 4 bytes of outbound p2p data

To learn more about these events refer to the paragraph 2.4 of the guide I've pointed above.
